src/sagas/dependency.saga.js

```diff
@@ -0,0 +1,325 @@
+  yield put(
+    updateNotification(notificationId, { title: 'Deleting dependencies...' })
+  );
+
+  const projectPath = yield select(getPathForProjectId, projectId);
+  try {
+    yield call(uninstallDependencies, projectPath, dependencies);
+    yield put(completeNotification(notificationId));
+    yield put(deleteDependenciesFinish(projectId, dependencies));
+  } catch (err) {
+    yield put(failNotification(notificationId, err));
+    yield put(deleteDependenciesError(projectId, dependencies));
+  }
+}
+
+function requestPackageJsonLockForProjectId(projectId) {
+  if (lock[projectId]) return false;
+
+  lock[projectId] = true;
+  return true;
+}
+
+function* enqueue({ action, projectId, dependencyName, version }) {
+  // If a package.json-locking function is already underway, queue
+  // this action up to be fired next in series. If there are already
+  // queued actions of the same type, append this action's information
+  // to that action so they can all be processed as a single command.
+  const projectQueue = queue[projectId] || [];
+
+  let queueIndex = projectQueue.findIndex(q => q.action === action);
+  const canCompound = queueIndex !== -1;
+
+  if (!canCompound) {
+    const notificationId = `${projectId}-queue-${Date.now()}`;
+    projectQueue.push({
+      action: action,
+      notificationId,
+      dependencies: [],
+    });
+    queueIndex = projectQueue.length - 1;
+    yield put(
+      showNotification(notificationId, {
+        title: `Dependencies queued for ${action}...`,
+        message: `${dependencyName}${version ? '@' + version : ''}`,
+      })
+    );
+  }
+
+  projectQueue[queueIndex].dependencies.push({
+    dependencyName,
+    version,
+  });
+  queue[projectId] = projectQueue;
+
+  if (canCompound) {
+    const queueAction = projectQueue[queueIndex];
+    yield put(
+      updateNotification(queueAction.notificationId, {
+        message: queueAction.dependencies
+          .map(
+            dependency =>
+              `${dependency.dependencyName}${
+                dependency.version ? '@' + dependency.version : ''
+              }`
+          )
+          .join(', '),
+      })
+    );
+  }
+}
+
+function* dequeue({ projectId }) {
+  console.log('lock:', lock);
+  console.log('queue:', queue);
+
+  // As soon a package.json-locking function completes, unlock the
+  // corresponding package.json.
+  lock[projectId] = false;
+
+  // Attempt to remove the next set of pending dependencies for that
+  // project from the queue and dispatch an action to run them.
+  if (queue[projectId] && queue[projectId].length > 0) {
+    const nextAction = queue[projectId].shift();
+    const actionCreator =
+      nextAction.action === 'install'
+        ? addDependenciesStart
+        : deleteDependenciesStart;
+    yield put(
+      actionCreator(
+        projectId,
+        nextAction.notificationId,
+        nextAction.dependencies
+      )
+    );
+  }
+}
+
+// $FlowFixMe
```

Comment:
What's Flow complaining about?

Reply:
Generators are supposed to be typed as function* foo()<Yield, Return, Next> { /*...*/ }, but I'm not really sure how to interpret any of those three types in the context of redux-saga - in this case it complains about `next()` being the wrong type. Feel free to have a hack at typing that, I'm out of my wheelhouse there 😆

Reply:
Ah geez, yeah, I'm reasonably comfortable with Flow but I've never used it with generators.

No worries! If @mik639 has any ideas, we can address it in a subsequent PR. Otherwise this isn't high on my priority list.